src/LandIce/evaluators/LandIce_BasalMeltRate_Def.hpp

```diff
@@ -107,55 +120,173 @@ postRegistrationSetup(typename Traits::SetupData d, PHX::FieldManager<Traits>& f
+                                                                        ScalarT(0.5 + 0.5* basal_reg_coeff * diffEnthalpy);
+                                                                  //    ScalarT(0.5 + 0.5* (0.5-0.5*std::pow(1-basal_reg_coeff * diffEnthalpy,2)));
+
+    //mstar, [W m^{-2}] = [Pa m s^{-1}]: basal latent heat in temperate ice
+    ScalarT mstar = geoFluxHeat(sideSet_idx,node);
+    for (int dim = 0; dim < vecDimFO; dim++)
+      mstar += 1000./scyr * beta(sideSet_idx,node) * velocity(sideSet_idx,node,dim) * velocity(sideSet_idx,node,dim);
+
+    double dTdz_melting = beta_p * rho_i * g;
+    mstar += 1e-3* k_i * dTdz_melting;
+
+    //enthalpyBasalFlux(cell,side,node) =  (basal_reg_scale-1) *mstar + 1e-3*k_i*dTdz_melting;
+    enthalpyBasalFlux(sideSet_idx,node) =  (basal_reg_scale-1) *mstar + 1e-3*k_i*dTdz_melting;
+
+    ScalarT basal_water_flux = scyr * k_0 * (rho_w - rho_i) * g / eta_w * pow(phi(sideSet_idx,node),alpha_om); //[m yr^{-1}]
+    ScalarT melting = scyr * basal_reg_scale * mstar / (L*rho_i); //[m yr^{-1}]
+    // basalVertVelocity(cell,side,node) =  - melting /(1 - rho_w/rho_i*deviceMin<ScalarT>(phi(sideSet_idx,node),0.5)) -  basal_water_flux;
+    basalVertVelocity(sideSet_idx,node) =  - melting /(1 - rho_w/rho_i*deviceMin<ScalarT>(phi(sideSet_idx,node),0.5)) -  basal_water_flux;
+  }
+
+}
+
+template<typename EvalT, typename Traits, typename VelocityST, typename MeltEnthST>
+KOKKOS_INLINE_FUNCTION
+void BasalMeltRate<EvalT,Traits,VelocityST,MeltEnthST>::
+operator() (const Basal_Melt_Rate_Tag& tag, const int& sideSet_idx) const {
+
+  const int dim = nodal ? numSideNodes : numSideQPs;
+
+  // Get the local data of side and cell
+  const int cell = sideSet.elem_LID(sideSet_idx);
+  const int side = sideSet.side_local_id(sideSet_idx);
+
+  for (int node = 0; node < dim; ++node) {
+    //always in presence of water on shelves (assuming that beta==0 <==> on shelves)
+    bool isThereWaterHere = isThereWater || (beta(cell,side,node) == 0.0);
+    ScalarT diffEnthalpy = Enthalpy(cell,side,node) - EnthalpyHs(cell,side,node);
+    ScalarT basal_reg_scale = (diffEnthalpy > 0 || !isThereWaterHere) ?  ScalarT(0.5 + 0.5*tanh(basal_reg_coeff * diffEnthalpy)) :
+                                                                        ScalarT(0.5 + 0.5* basal_reg_coeff * diffEnthalpy);
+                                                                  //    ScalarT(0.5 + 0.5* (0.5-0.5*std::pow(1-basal_reg_coeff * diffEnthalpy,2)));
+
+    //mstar, [W m^{-2}] = [Pa m s^{-1}]: basal latent heat in temperate ice
+    ScalarT mstar = geoFluxHeat(cell,side,node);
+    for (int dim = 0; dim < vecDimFO; dim++)
+      mstar += 1000./scyr * beta(cell,side,node) * velocity(cell,side,node,dim) * velocity(cell,side,node,dim);
+
+    double dTdz_melting = beta_p * rho_i * g;
+    mstar += 1e-3* k_i * dTdz_melting;
```

Comment:
It seems to me that `1e-3*k_i*dTdx_melting` could be precomputed at construction time, no?

Reply:
This was mentioned for the volume evaluator PR as well. In that PR, I changed all of the scaling constants like this to be precomputed and it had no measurable effect on the performance. My understanding is that CUDA is smart enough to optimize these kinds of things away. I think having them like this is more readable but it might not be portable. I'll change these to be precomputed for the commit I'm working on but I think it might not be necessary.

Reply:
On CUDA, flops are basically free, but all of these things might add up in CPU builds.

Reply:
This might get optimized out on a CPU build too. But I agree, it's probably better just to pull this out of the kernel just in case.